src/Orleans.Runtime/Cancellation/GrainCallCancellationManager.cs

```diff
@@ -0,0 +1,329 @@
+        ActivationDirectory activationDirectory,
+        IClusterMembershipService clusterMembershipService) : base(Constants.CancellationManagerType, localSiloDetails.SiloAddress, loggerFactory)
+    {
+        _serviceProvider = serviceProvider;
+        _logger = loggerFactory.CreateLogger<GrainCallCancellationManager>();
+        _catalog = catalog;
+        _activationDirectory = activationDirectory;
+        _clusterMembershipService = clusterMembershipService;
+        _catalog.RegisterSystemTarget(this);
+
+        {
+            using var _ = new ExecutionContextSuppressor();
+            _membershipUpdatesTask = Task.Factory.StartNew(
+                state => ((GrainCallCancellationManager)state!).ProcessMembershipUpdates(),
+                this,
+                CancellationToken.None,
+                TaskCreationOptions.None,
+                WorkItemGroup.TaskScheduler).Unwrap();
+            _membershipUpdatesTask.Ignore();
+        }
+    }
+
+    private IInternalGrainFactory GrainFactory => _grainFactory ??= _serviceProvider.GetRequiredService<IInternalGrainFactory>();
+
+    public ValueTask CancelCallsAsync(List<GrainCallCancellationRequest> cancellationRequests)
+    {
+        foreach (var request in cancellationRequests)
+        {
+            // Try to directly call the cancellation method locally
+            ValueTask cancellationTask;
+            if (_activationDirectory.FindTarget(request.TargetGrainId) is IGrainCallCancellationExtension extension)
+            {
+                cancellationTask = extension.CancelRequestAsync(request.SourceGrainId, request.MessageId);
+            }
+            else
+            {
+                // Fall back to a regular grain call.
+                cancellationTask = GrainFactory.GetGrain<IGrainCallCancellationExtension>(request.TargetGrainId).CancelRequestAsync(request.SourceGrainId, request.MessageId);
+            }
+
+            if (cancellationTask.IsCompletedSuccessfully)
+            {
+                // Observe the completed task.
+                cancellationTask.GetAwaiter().GetResult();
+            }
+            else
+            {
+                cancellationTask.AsTask().Ignore();
+            }
+        }
+
+        return ValueTask.CompletedTask;
+    }
+
+    public void SignalCancellation(SiloAddress? targetSilo, GrainId targetGrainId, GrainId sourceGrainId, CorrelationId messageId)
+    {
+        if (targetSilo is not null
+            && GetOrCreateWorker(targetSilo).Writer.TryWrite(new GrainCallCancellationRequest(targetGrainId, sourceGrainId, messageId)))
+        {
+            return;
+        }
+
+        var request = GrainFactory.GetGrain<IGrainCallCancellationExtension>(targetGrainId).CancelRequestAsync(sourceGrainId, messageId);
+        if (!request.IsCompleted)
+        {
+            request.AsTask().Ignore();
+        }
+    }
+
+    private async Task ProcessMembershipUpdates()
+    {
+        await Task.Yield();
```

Comment:
I think technically
`await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);`
is a little more efficient.